libnmstate/state.py

```diff
@@ -241,6 +241,47 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, other_state):
+        """
+        Merge config routes from other state to self.
+        Return a list of interface's name where route changed.
+        """
+        # Merge other_routes
+        self_iface_route_sets = defaultdict(set)
+        other_iface_route_sets = defaultdict(set)
+
+        for route in self._config_routes:
+            if route.get(Route.STATE) != Route.STATE_ABSENT:
+                if Route.NEXT_HOP_INTERFACE in route:
+                    self_iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                        RouteEntry(route))
+
+        for route in other_state._config_routes:
+            # The other_state(running state) does not have any absent
+            # route entry and all have NEXT_HOP_INTERFACE
+            iface_name = route[Route.NEXT_HOP_INTERFACE]
+            self_iface_route_sets[iface_name].add(RouteEntry(route))
+            # Adding to other_iface_route_sets so that we could
+            # known which interface changed.
+            other_iface_route_sets[iface_name].add(RouteEntry(route))
+
+        merged_routes = []
+        for iface_name, route_set in six.viewitems(self_iface_route_sets):
+            merged_routes.extend(
+                list(route_obj.to_dict() for route_obj in route_set))
+
+        self._config_routes = merged_routes
+        self._config_iface_routes = State._index_routes_by_iface(
+            self._config_routes)
+
+        changed_interface = []
```

Comment:
When there is a method that changes the internal state (like this merge one), it is preferred not to return anything except perhaps `self` or the data that was just processed (even that I usually do not like).
The reason for it usually is that it is not expected.. it is kind of a surprise to the user of the method.

It is not a hard rule, but I found it useful to look for alternatives.
Could you please explain why you return it? Perhaps we can even delay the discussion of the why to the next change which will actually use it. Then we can look and see if there is any alternative or not.

Reply:
It's need when user only define routes without interfaces section entry.
Code removed.